subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c

```diff
@@ -0,0 +1,928 @@
+			p->message_id,
+			p->message.payload.len);
+
+		if (is_fota_active()) {
+			LOG_INF("Job in progress... skipping");
+			goto send_ack;
+		}
+
+		current_fota.mqtt_payload_size = p->message.payload.len + 1;
+		current_fota.mqtt_payload =
+			nrf_cloud_calloc(current_fota.mqtt_payload_size, 1);
+
+		if (!current_fota.mqtt_payload) {
+			LOG_ERR("Unable to allocate memory for job");
+			err = -ENOMEM;
+			goto send_ack;
+		}
+
+		ret = mqtt_readall_publish_payload(client_mqtt,
+				current_fota.mqtt_payload,
+				p->message.payload.len);
+		if (ret) {
+			LOG_ERR("Error reading MQTT payload: %d", ret);
+			err = ret;
+			goto send_ack;
+		}
+
+		ret = parse_job(&current_fota);
+		if (ret) {
+			err = ret;
+			goto send_ack;
+		}
+
+		start = true;
+send_ack:
+		if (p->message.topic.qos == MQTT_QOS_0_AT_MOST_ONCE) {
+			LOG_DBG("No ack required");
+		} else {
+			ret = mqtt_publish_qos1_ack(client_mqtt, &ack);
+			if (ret) {
+				LOG_ERR("MQTT ACK failed: %d", ret);
+			}
+		}
+
+		if (start) {
+			ret = start_job(&current_fota);
+			(void)send_job_update(&current_fota);
+			if (ret) {
+				cleanup_job(&current_fota);
+				ret = 0;
+			}
+		}
+
+		return (err ? err : ret);
+	}
+	case MQTT_EVT_SUBACK:
+	{
+		if (evt->param.suback.message_id !=
+		    NRF_CLOUD_FOTA_SUBSCRIBE_ID) {
+			return 1;
+		}
+		LOG_DBG("MQTT_EVT_SUBACK");
+
+		nrf_cloud_fota_update_check();
+
+		break;
+	}
+	case MQTT_EVT_UNSUBACK:
+	{
+		if (evt->param.unsuback.message_id !=
+		    NRF_CLOUD_FOTA_SUBSCRIBE_ID) {
+			return 1;
+		}
+		LOG_DBG("MQTT_EVT_UNSUBACK");
+		break;
+	}
+	case MQTT_EVT_PUBACK:
+	{
+		bool do_update_check = false;
+
+		if ((evt->param.puback.message_id !=
+		    NRF_CLOUD_FOTA_UPDATE_ID) &&
+		    (evt->param.puback.message_id !=
+		    NRF_CLOUD_FOTA_REQUEST_ID)) {
+			return 1;
+		}
+
+		LOG_DBG("MQTT_EVT_PUBACK: msg id %d",
+			evt->param.puback.message_id);
+
+		if (evt->param.puback.message_id ==
+		    NRF_CLOUD_FOTA_REQUEST_ID) {
+			/* Nothing to do */
+			break;
+		}
+
+		do_update_check = !is_fota_active();
+
+		switch (saved_job.validate) {
+		case NRF_FOTA_VALIDATE_PASS:
+		case NRF_FOTA_VALIDATE_UNKNOWN:
+		case NRF_FOTA_VALIDATE_FAIL:
+			save_validate_status(saved_job.id, saved_job.type,
+					NRF_FOTA_VALIDATE_DONE);
+			break;
+		case NRF_FOTA_VALIDATE_PENDING:
+			/* this event should cause reboot */
+			send_event(NRF_FOTA_EVT_DONE, &current_fota);
+			cleanup_job(&current_fota);
+			do_update_check = false;
+			break;
+		default:
+			break;
+		}
+
+		if (do_update_check) {
+			/* TODO:
+			 * This shouldn't be needed as the
+			 * backend will send jobs automatically
+			 * when one completes. Verify backend
+			 * implementation works as expected and
+			 * then remove this.
+			 */
+			//nrf_cloud_fota_update_check();
+		}
+
+		break;
+	}
+	case MQTT_EVT_CONNACK:
+	case MQTT_EVT_DISCONNECT:
+	case MQTT_EVT_PUBREC:
+	case MQTT_EVT_PUBREL:
+	case MQTT_EVT_PUBCOMP:
+	case MQTT_EVT_PINGRESP:
```

Comment:
Perhaps add a default case too, in the unlikely event that a new MQTT event is added upstream